src/Sarif.Viewer.VisualStudio/Telemetry/Events.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Sarif.Viewer.Telemetry
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Contains the methods for firing telemetry events from the SARIF viewer extension.
+    /// </summary>
+    /// <remarks>
+    /// The documentation in this class is super useful for understanding where this telemetry is fired.
+    /// The documentation should also contain "why" the telemetry is being fired, i.e., why is this data needed?
+    /// Also, by having individual methods, versus calling application insights directly from the code,
+    /// it allows us to easily locate the code that is firing the telemetry.
+    /// </remarks>
+    internal class Events
+    {
+        /// <summary>
+        /// Does nothing.
+        /// </summary>
+        /// <remarks>
+        /// The reason we call track event directly here is that
+        /// we are in the middle of MEF composition\creation <see cref="TelemetryProvider.TelemetryProviderService"/>, if we called
+        /// <see cref="TelemetryProvider.TrackEvent(string)"/> that would end up being a recursive MEF composition call (and a
+        /// MEF composition exception). This method is purely used for event naming purposes.
+        /// </remarks>
+        public static void ExtensionLoaded() => throw new NotImplementedException();
```

Comment:
>throw new NotImplementedException(); [](start = 48, length = 36)

BUG: Presumably you meant to call `TelemetryClient.TrackEvent` directly here, and the comment suggests.

Reply:
No, actually I did not... This should not be called, it is here for naming purposes only (see comment)

Reply:
No, actually I did not... This should not be called, it is here for naming purposes only (see comment)

---
In reply to: [505751525](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505751525) [](ancestors = 505751525)

Reply:
The comment says "The reason we call track event directly here..." but you don't call track event here at all.

---
In reply to: [505752538](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505752538) [](ancestors = 505752538)

Reply:
Ok, I see that the call is in `TelemetryProvider` itself. I can't see creating a throw-only method just so you can use its _name_ somewhere else (`nameof(ExtensionLoaded)`).

If you are concerned about hard-coding a name into TelemetryProvider, and then somebody accidentally creates a "real" event with the same name in this class, then just defined
```
internal const string ExtensionLoaded = nameof(ExtensionLoaded);
```

---
In reply to: [505753641](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505753641) [](ancestors = 505753641,505752538)

Reply:
I'll add the work "don't"

---
In reply to: [505753641](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505753641) [](ancestors = 505753641,505752538)

Reply:
IMO adding the word "don't" to a comment that explains why you should never call a method that always throws won't do. If you need a string, declare a string. You don't need a method.

---
In reply to: [505756813](https://github.com/microsoft/sarif-visualstudio-extension/pull/226#discussion_r505756813) [](ancestors = 505756813,505753641,505752538)